modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -817,5 +818,1120 @@ int HashTSDFVolumeCPU::getVisibleBlocks(int currFrameId, int frameThreshold) con
+
+        for (int i = range.start; i < range.end; ++i)
+        {
+            Vec3i tsdf_idx = _totalVolUnits[i];
+
+            VolumeIndex idx = _indexes.find_Volume(tsdf_idx);
+            if (idx < 0 || idx == _lastVolIndex - 1) return;
+
+            Point3f volumeUnitPos = volumeUnitIdxToVolume(poses.at<Vec3i>(idx, 0));
+            Point3f volUnitInCamSpace = vol2cam * volumeUnitPos;
+
+            if (volUnitInCamSpace.z < 0 || volUnitInCamSpace.z > truncateThreshold)
+            {
+                _indexes.updateActive(tsdf_idx, 0);
+                return;
+            }
+            Point2f cameraPoint = proj(volUnitInCamSpace);
+            if (cameraPoint.x >= 0 && cameraPoint.y >= 0 && cameraPoint.x < depth.cols && cameraPoint.y < depth.rows)
+            {
+                assert(idx == _lastVolIndex - 1);
+                lastVisibleIndexes.at<int>(idx, 0) = frameId;
+                _indexes.update(tsdf_idx, 1, frameId);
+            }
+        }
+    };
+    parallel_for_(_inFrustumRange, markActivities);
+    //markActivities(_inFrustumRange);
+
+    Vec6f newParams((float)depth.rows, (float)depth.cols,
+        intrinsics.fx, intrinsics.fy,
+        intrinsics.cx, intrinsics.cy);
+    if (!(frameParams == newParams))
+    {
+        frameParams = newParams;
+        pixNorms = preCalculationPixNorm(depth, intrinsics);
+        Vec2f fxy(intrinsics.fx, intrinsics.fy), cxy(intrinsics.cx, intrinsics.cy);
+        _pixNorms = preCalculationPixNormGPU(depth.rows, depth.cols, fxy, cxy);
+    }
+
+    //! Integrate the correct volumeUnits
+    /*
+    auto Integrate = [&](const Range& range) {
+        for (int i = range.start; i < range.end; i++)
+        {
+            Vec3i tsdf_idx = _totalVolUnits[i];
+            VolumeIndex idx = _indexes.find_Volume(tsdf_idx);
+            if (idx < 0 || idx == _lastVolIndex - 1) return;
+
+            bool _isActive = _indexes.getActive(tsdf_idx);
+
+            if (_isActive)
+            {
+                //! The volume unit should already be added into the Volume from the allocator
+                Matx44f _pose = poses.at<Matx44f>(idx, 0);
+
+                integrateVolumeUnit(truncDist, voxelSize, maxWeight, _pose,
+                    Point3i(volumeUnitResolution, volumeUnitResolution, volumeUnitResolution), volStrides, depth,
+                    depthFactor, cameraPose, intrinsics, pixNorms, _volUnitsData.row(idx));
+
+                //integrateVolumeUnitGPU(depth, depthFactor, _pose, intrinsics, idx);
+                //! Ensure all active volumeUnits are set to inactive for next integration
+                _indexes.updateActive(tsdf_idx, 0);
+            }
+        }
+    };
+    */
+    //parallel_for_(Range(0, (int)_totalVolUnits.size()), Integrate );
+    //Integrate(Range(0, (int)_totalVolUnits.size()));
+
+    integrateAllVolumeUnitsGPU(depth, depthFactor, intrinsics);
+
+}
+
+cv::Vec3i HashTSDFVolumeGPU::volumeToVolumeUnitIdx(const cv::Point3f& p) const
```

Comment:
I think this function (as well as 3 other below) should be moved to `tsdf_functions.hpp` since they are used in both GPU and CPU versions.

Reply:
Disregard that, these functions will be removed as soon as we have a proper GPU version of fetchPoints/Normals.